### PR TITLE
Add Mysql as a checkpoint store, add tests, and add example consumer

### PR DESCRIPTION
**Pull request by fuziontech**
Definitely helpful over here, might be helpful for someone else!

### CODE REVIEW

**Comment by harlow** [comment]
Hi @fuziontech thanks for the PR! Give me a day or two to test this w/ Mysql (I currently don't have running locally). Once confirmed, I'll get this merged asap. 👍

File: examples/consumer/cp-mysql/main.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+
+	go func() {
+		<-signals
+		cancel()
+	}()
+
+	// scan stream
+	err = c.Scan(ctx, func(r *consumer.Record) consumer.ScanStatus {
```
Comment (harlow) [inline examples/consumer/cp-mysql/main.go]: > examples/consumer/cp-mysql/main.go:56:45: undefined: consumer.ScanStatus

looks like tests are failing here. The `consumer.ScanStatus` concept has been removed from the repo, and replaced with a slightly different interface (`error` to stop, or `nil` to continue scanning):

```go
err = c.Scan(context.TODO(), func(r *consumer.Record) error {
  fmt.Println(string(r.Data))
  return nil // continue scanning
})
```

Comment (fuziontech) [reply examples/consumer/cp-mysql/main.go]: Yup! Good catch. Updated it to the newer interface. Thanks for the quick feedback!

File: checkpoint/mysql/mysql.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+func New(appName, tableName, connectionStr string, opts ...Option) (*Checkpoint, error) {
+	if appName == "" {
+		return nil, errors.New("application name not defined")
+	}
+
+	if tableName == "" {
+		return nil, errors.New("table name not defined")
+	}
+
+	conn, err := sql.Open("mysql", connectionStr)
```
Comment (harlow) [inline checkpoint/mysql/mysql.go]: if we passed in the `conn` to `New` would the Postgres and Mysql implementation be interchangeable? Nothing else jumps out to me as different from the two checking point implementations

File: checkpoint/mysql/mysql.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
+package mysql
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"sync"
+	"time" // this is the mysql package so it makes sense to be here
```
Comment (harlow) [inline checkpoint/mysql/mysql.go]: can we move this comment down, not sure it make sense next to "time"

File: checkpoint/mysql/mysql.go (hunk 2 above)
Comment (fuziontech) [reply checkpoint/mysql/mysql.go]: Yeah, the duplicating everything does have a smell to it. The only other thing that changed is the upsert query in the save function. The select function changed a bit too, just `$1` syntax to `?` for params.
For mysql:
```go
upsertCheckpoint := fmt.Sprintf(`REPLACE INTO %s (namespace, shard_id, sequence_number) VALUES (?, ?, ?)`, c.tableName)
```
vs for PG:
```go
upsertCheckpoint := fmt.Sprintf(`INSERT INTO %s (namespace, shard_id, sequence_number)
				VALUES($1, $2, $3)
				ON CONFLICT (namespace, shard_id)
				DO 
				UPDATE 
				SET sequence_number= $3;`, c.tableName)
```

Could definitely refactor to have a more generic sql module that msql and pg modules are based off of where you just provide the conn string and upsert statement. I felt that was a bit invasive for my first PR on this though 😛

File: checkpoint/mysql/mysql.go (hunk 3 above)
Comment (fuziontech) [reply checkpoint/mysql/mysql.go]: We can just drop it altogether. That was from the copy pasta from postgres implementation.

File: checkpoint/mysql/mysql.go (hunk 2 above)
Comment (harlow) [reply checkpoint/mysql/mysql.go]: got it. ok sounds good. i'm on board w/ keeping the duplication for the time being.

there might also be a world where we have templates for each of the different dbs 

```go
var upsertTemplates = map[string]string{
	"postgres": `
		INSERT INTO %s (namespace, shard_id, sequence_number)
		VALUES($1, $2, $3)
		ON CONFLICT (namespace, shard_id)
		DO UPDATE SET sequence_number= $3;
	`,
	"mysql": `
		REPLACE INTO %s (namespace, shard_id, sequence_number)
		VALUES (?, ?, ?)
	`,
}
```

**Pull request merged**